### PR TITLE
Fixing explore actions & slice controller interactions

### PR DESCRIPTION
**Pull request by mistercrunch**
The idea here is `controllerInterface`, a common interface for slices to interact with their context, or controller (explore, dashboard or standalone implement this interface).

For instance, when a slice is done loading, it calls it's controller `done` method, which will trigger different things depending on the context.

This allows for moving logic specific to dashboards & explore back into those objects, away from the Slice object itself.

### CODE REVIEW

File: caravel/assets/javascripts/modules/caravel.js [hunk 1]
```diff
@@ -76,16 +77,13 @@ const px = function () {
...
+      querystring() {
         const parser = document.createElement('a');
         parser.href = data.json_endpoint;
-        if (dashboard !== undefined) {
-          let flts = '';
-          if (newParams.extraFilters !== false) {
-            flts = dashboard.effectiveExtraFilters(sliceId);
-            flts = encodeURIComponent(JSON.stringify(flts));
-          }
+        if (controller.type === 'dashboard') {
```
Comment (prokh) [inline caravel/assets/javascripts/modules/caravel.js]: 1. origJsonEndpoint with the params for the chart is only used in the dashboard to be able to remove the extra_filters. Hence, I would save it in the dashboard.
2. The form input with the params for the chart is only used in explore.
3. In the default case, e.g. standalone, the query params are from the slide itself. 
Thus, I would put the query string generation in the controller interface in use the default case when it is not implemented.

Comment (prokh) [inline caravel/assets/javascripts/modules/caravel.js]: Would you please describe the interface with a few comments. e.g. describe the case when each function is called.

File: caravel/assets/javascripts/explore/explore.jsx [hunk 2]
```diff
@@ -345,23 +356,40 @@ function initComponents() {
...
+
 $(document).ready(function () {
   const data = $('.slice').data('slice');
 
   initExploreView();
 
-  slice = px.Slice(data);
+  slice = px.Slice(data, exploreController);
 
   $('.slice').data('slice', slice);
```
Comment (mistercrunch) [inline caravel/assets/javascripts/explore/explore.jsx]: good catch, I'm removing it and making sure things still work. We're re-writing the explore view to react so I don't want to go too far in the refactoring, but that's an easy win.

File: caravel/assets/javascripts/modules/caravel.js (hunk 1 above)
Comment (mistercrunch) [inline caravel/assets/javascripts/modules/caravel.js]: thanks for the review, I know there's much more refactoring to do here to move all the logic in the right places, I want to merge this quickly-ish to fix the bug and we can see how much more refactoring we want to do here in the future but most likely not much knowing that we're re-writing the whole explore view...

**Comment by ascott** [comment]
thanks for this fix! 🎉

**Pull request merged**